File: backend/app/services/reservation_service.py
"""Equivalente a services/reservation-service.ts."""
from datetime import datetime, timedelta, timezone

from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

from app.errors import BusinessError, ResourceNotFoundError
from app.lock import LockOptions, RedisLockManager
from app.logging_config import log
from app.mappers import to_reservation_response
from app.repositories import events as events_repo
from app.repositories import reservations as reservations_repo
from app.repositories import tickets as tickets_repo
from app.repositories.reservations import ReservationView
from app.schemas import CreateReservationRequest, ReservationResponse
from app.token import generate_reservation_token

# Mesmos tempos do original (tryLock(wait, lease)).
CREATE_LOCK = LockOptions(wait_ms=5_000, lease_ms=10_000)
RELEASE_LOCK = LockOptions(wait_ms=5_000, lease_ms=5_000)
CONFIRM_EXTENSION_MINUTES = 5


def _now() -> datetime:
    """"Agora", em UTC, sem tzinfo. As colunas TIMESTAMP do Postgres são "sem fuso" e este
    sistema sempre grava/lê UTC nelas (mesma convenção do db.ts original) — manter os `datetime`
    da aplicação *naive* evita misturar aware/naive nas comparações e no driver asyncpg."""
    return datetime.now(timezone.utc).replace(tzinfo=None)


def is_reservation_expired(reservation: ReservationView, now: datetime) -> bool:
    """Equivale a `Reservation.isExpired()`: expirada quando "agora" está estritamente depois de expiresAt."""
    return now > reservation.expires_at


async def release_active_reservation(session: AsyncSession, reservation: ReservationView) -> bool:
    """Devolve o assento e a vaga do evento e cancela a reserva — se ela ainda estiver ACTIVE.

    Deve rodar dentro de uma transação. Ordem fixa de locks (evento -> assento -> reserva), a mesma
    usada em create_reservation e no pagamento, para que transações concorrentes nunca se travem em
    ciclo. A reserva é relida já com lock: quem chamou pode ter uma versão desatualizada (ex.: um
    pagamento concluiu enquanto o job de expiração esperava).
    """
    await events_repo.find_by_id_for_update(session, reservation.event_id)
    await tickets_repo.find_by_id_for_update(session, reservation.ticket_id)
    current = await reservations_repo.find_by_id_for_update(session, reservation.id)
    if current is None or current.status != "ACTIVE":
        return False

    await tickets_repo.update_status(session, reservation.ticket_id, "AVAILABLE")
    await events_repo.increment_available(session, reservation.event_id)
    await reservations_repo.update_status(session, reservation.id, "CANCELLED")
    return True


class ReservationService:
    def __init__(
        self,
        session_factory: async_sessionmaker[AsyncSession],
        locks: RedisLockManager,
        reservation_timeout_minutes: int,
    ) -> None:
        self._session_factory = session_factory
        self._locks = locks
        self._timeout_minutes = reservation_timeout_minutes

    async def get_reservation_by_token(self, token: str) -> ReservationView:
        async with self._session_factory() as session:
            reservation = await reservations_repo.find_by_token(session, token)
        if reservation is None:
            raise ResourceNotFoundError("Reservation not found")
        return reservation

    async def create_reservation(self, request: CreateReservationRequest) -> ReservationResponse:
        lock_key = f"event:{request.eventId}:ticket:{request.seatNumber}"

        async def do_create() -> ReservationResponse:
            async with self._session_factory() as session:
                async with session.begin():
                    event = await events_repo.find_by_id_for_update(session, request.eventId)
                    if event is None:
                        raise ResourceNotFoundError("Event not found")
                    if event.available_tickets <= 0:
                        raise BusinessError("No tickets available for this event")

                    ticket = await tickets_repo.find_by_event_and_seat_for_update(session, request.eventId, request.seatNumber)
                    if ticket is None:
                        raise ResourceNotFoundError("Seat not found")
                    if ticket.status != "AVAILABLE":
                        raise BusinessError("Seat is not available")

                    if await reservations_repo.exists_active_for_ticket(session, ticket.id):
                        raise BusinessError("Seat already has an active reservation")

                    now = _now()
                    await tickets_repo.update_status(session, ticket.id, "RESERVED")
                    await events_repo.decrement_available(session, event.id)
                    reservation = await reservations_repo.insert(
                        session,
                        reservation_token=generate_reservation_token(),
                        event_id=event.id,
                        ticket_id=ticket.id,
                        user_id=request.userId,
                        user_email=request.userEmail,
                        expires_at=now + timedelta(minutes=self._timeout_minutes),
                    )
                    log.info("Reservation created successfully: %s", reservation.reservation_token)
                    return to_reservation_response(reservation, reservation.event_name, reservation.seat_number, now)

        # Ordem: lock distribuído -> transação -> COMMIT -> libera o lock.
        result = await self._locks.try_with_lock(lock_key, CREATE_LOCK, do_create)
        if not result.acquired:
            raise BusinessError("Could not acquire lock for seat reservation")
        assert result.value is not None
        return result.value

    async def release_reservation(self, token: str) -> None:
        reservation = await self.get_reservation_by_token(token)

        async def do_release() -> bool:
            async with self._session_factory() as session:
                async with session.begin():
                    return await release_active_reservation(session, reservation)

        result = await self._locks.try_with_lock(f"reservation:{token}", RELEASE_LOCK, do_release)
        # Igual ao original: não conseguir o lock não é erro (quem o segura está mexendo na reserva).
        if not result.acquired:
            log.warning("Reservation %s not released: lock busy", token)
        elif result.value:
            log.info("Reservation released: %s", token)

    async def expire_reservations(self) -> int:
        async with self._session_factory() as session:
            expired = await reservations_repo.find_expired_active(session, _now())
        log.info("Found %d expired reservations to process", len(expired))

        released = 0
        for reservation in expired:
            try:
                await self.release_reservation(reservation.reservation_token)
                released += 1
            except Exception as error:  # noqa: BLE001
                log.error("Error processing expired reservation: %s (%s)", reservation.reservation_token, error)
        return released

    async def confirm_reservation(self, token: str) -> ReservationResponse:
        reservation = await self.get_reservation_by_token(token)
        now = _now()

        if reservation.status != "ACTIVE":
            raise BusinessError("Reservation is not active")
        if is_reservation_expired(reservation, now):
            raise BusinessError("Reservation has expired")

        # Comportamento herdado: a nova expiração é "agora + 5 min" (pode até ser menor que a atual).
        expires_at = now + timedelta(minutes=CONFIRM_EXTENSION_MINUTES)
        async with self._session_factory() as session:
            async with session.begin():
                updated = await reservations_repo.update_expires_at(session, reservation.id, expires_at)
        if not updated:
            raise BusinessError("Reservation is not active")

        reservation.expires_at = expires_at
        return to_reservation_response(reservation, reservation.event_name, reservation.seat_number, now)

    async def get_reservation_response_by_token(self, token: str) -> ReservationResponse:
        reservation = await self.get_reservation_by_token(token)
        return to_reservation_response(reservation, reservation.event_name, reservation.seat_number, _now())

    async def is_reservation_valid(self, token: str) -> bool:
        try:
            reservation = await self.get_reservation_by_token(token)
        except ResourceNotFoundError:
            return False
        return reservation.status == "ACTIVE" and not is_reservation_expired(reservation, _now())

File: src/lib/services/reservation-service.ts
import type { Database, Queryable } from '../db-types';
import { toReservationResponse, type ReservationResponse } from '../dto';
import { BusinessError, ResourceNotFoundError } from '../errors';
import type { LockManager } from '../lock';
import { log } from '../logger';
import type { Repositories } from '../repositories';
import { isReservationExpired, type Reservation } from '../types';
import type { CreateReservationRequest } from '../validation';

export interface ReservationService {
  createReservation(request: CreateReservationRequest): Promise<ReservationResponse>;
  getReservationByToken(token: string): Promise<Reservation>;
  getReservationResponseByToken(token: string): Promise<ReservationResponse>;
  confirmReservation(token: string): Promise<ReservationResponse>;
  /** Libera todas as reservas ACTIVE vencidas; devolve quantas foram liberadas. */
  expireReservations(): Promise<number>;
  releaseReservation(token: string): Promise<void>;
  isReservationValid(token: string): Promise<boolean>;
}

export interface ReservationServiceDeps {
  db: Database;
  repos: Repositories;
  locks: LockManager;
  generateToken: () => string;
  now: () => Date;
  reservationTimeoutMinutes: number;
}

// Mesmos tempos do original (tryLock(wait, lease)).
const CREATE_LOCK = { waitMs: 5_000, leaseMs: 10_000 };
const RELEASE_LOCK = { waitMs: 5_000, leaseMs: 5_000 };
const CONFIRM_EXTENSION_MINUTES = 5;

/**
 * Devolve o assento e a vaga do evento e cancela a reserva — se ela ainda estiver ACTIVE.
 * Deve rodar dentro de uma transação. Ordem fixa de locks (evento → assento → reserva), a mesma
 * usada em createReservation e no pagamento, para que transações concorrentes nunca se travem em ciclo.
 * A reserva é relida já com lock: quem chamou pode ter uma versão desatualizada (ex.: um pagamento
 * concluiu enquanto o job de expiração esperava).
 */
export async function releaseActiveReservation(repos: Repositories, tx: Queryable, reservation: Reservation): Promise<boolean> {
  await repos.events.findByIdForUpdate(tx, reservation.eventId);
  await repos.tickets.findByIdForUpdate(tx, reservation.ticketId);
  const current = await repos.reservations.findByIdForUpdate(tx, reservation.id);
  if (!current || current.status !== 'ACTIVE') return false;

  await repos.tickets.updateStatus(tx, reservation.ticketId, 'AVAILABLE');
  await repos.events.incrementAvailable(tx, reservation.eventId);
  await repos.reservations.updateStatus(tx, reservation.id, 'CANCELLED');
  return true;
}

export function createReservationService(deps: ReservationServiceDeps): ReservationService {
  const { db, repos, locks } = deps;

  async function getReservationByToken(token: string): Promise<Reservation> {
    const reservation = await repos.reservations.findByToken(db.pool, token);
    if (!reservation) throw new ResourceNotFoundError('Reservation not found');
    return reservation;
  }

  async function createReservation(request: CreateReservationRequest): Promise<ReservationResponse> {
    const lockKey = `event:${request.eventId}:ticket:${request.seatNumber}`;

    // Ordem: lock distribuído → transação → COMMIT → libera o lock. (No original o lock era solto
    // antes do commit, porque o @Transactional envolvia o método inteiro.)
    const result = await locks.tryWithLock(lockKey, CREATE_LOCK, () =>
      db.transaction(async (tx) => {
        const event = await repos.events.findByIdForUpdate(tx, request.eventId);
        if (!event) throw new ResourceNotFoundError('Event not found');

        if (event.availableTickets <= 0) throw new BusinessError('No tickets available for this event');

        const ticket = await repos.tickets.findByEventAndSeatForUpdate(tx, request.eventId, request.seatNumber);
        if (!ticket) throw new ResourceNotFoundError('Seat not found');

        if (ticket.status !== 'AVAILABLE') throw new BusinessError('Seat is not available');

        if (await repos.reservations.existsActiveForTicket(tx, ticket.id)) {
          throw new BusinessError('Seat already has an active reservation');
        }

        const now = deps.now();
        await repos.tickets.updateStatus(tx, ticket.id, 'RESERVED');
        await repos.events.decrementAvailable(tx, event.id);
        const reservation = await repos.reservations.insert(tx, {
          reservationToken: deps.generateToken(),
          eventId: event.id,
          ticketId: ticket.id,
          userId: request.userId,
          userEmail: request.userEmail,
          expiresAt: new Date(now.getTime() + deps.reservationTimeoutMinutes * 60_000),
        });

        log.info(`Reservation created successfully: ${reservation.reservationToken}`);
        return toReservationResponse(reservation, now);
      }),
    );

    if (!result.acquired) throw new BusinessError('Could not acquire lock for seat reservation');
    return result.value;
  }

  async function releaseReservation(token: string): Promise<void> {
    const reservation = await getReservationByToken(token);

    const result = await locks.tryWithLock(`reservation:${token}`, RELEASE_LOCK, () =>
      db.transaction((tx) => releaseActiveReservation(repos, tx, reservation)),
    );

    // Igual ao original: não conseguir o lock não é erro (quem o segura está mexendo na reserva).
    if (!result.acquired) log.warn(`Reservation ${token} not released: lock busy`);
    else if (result.value) log.info(`Reservation released: ${token}`);
  }

  async function expireReservations(): Promise<number> {
    const expired = await repos.reservations.findExpiredActive(db.pool, deps.now());
    log.info(`Found ${expired.length} expired reservations to process`);

    let released = 0;
    for (const reservation of expired) {
      try {
        await releaseReservation(reservation.reservationToken);
        released++;
      } catch (error) {
        log.error(`Error processing expired reservation: ${reservation.reservationToken}`, error);
      }
    }
    return released;
  }

  async function confirmReservation(token: string): Promise<ReservationResponse> {
    const reservation = await getReservationByToken(token);
    const now = deps.now();

    if (reservation.status !== 'ACTIVE') throw new BusinessError('Reservation is not active');
    if (isReservationExpired(reservation, now)) throw new BusinessError('Reservation has expired');

    // Comportamento herdado: a nova expiração é "agora + 5 min" (pode até ser menor que a atual).
    const expiresAt = new Date(now.getTime() + CONFIRM_EXTENSION_MINUTES * 60_000);
    const updated = await repos.reservations.updateExpiresAt(db.pool, reservation.id, expiresAt);
    if (!updated) throw new BusinessError('Reservation is not active');

    return toReservationResponse({ ...reservation, expiresAt }, now);
  }

  async function getReservationResponseByToken(token: string): Promise<ReservationResponse> {
    return toReservationResponse(await getReservationByToken(token), deps.now());
  }

  async function isReservationValid(token: string): Promise<boolean> {
    try {
      const reservation = await getReservationByToken(token);
      return reservation.status === 'ACTIVE' && !isReservationExpired(reservation, deps.now());
    } catch (error) {
      if (error instanceof ResourceNotFoundError) return false;
      throw error;
    }
  }

  return {
    createReservation,
    getReservationByToken,
    getReservationResponseByToken,
    confirmReservation,
    expireReservations,
    releaseReservation,
    isReservationValid,
  };
}
